FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -26,17 +28,63 @@ public final class CoreDataFeedStore: FeedStore {
 			url: storeURL
 		)
 		context = container.newBackgroundContext()
+		cacheStore = ManagedCacheStore(context: context)
+		feedImageStore = ManagedFeedImageStore(context: context)
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [cacheStore] in
+			do {
+				if let cache = try cacheStore.cache() {
+					let localFeedImages = cache.feedImages.map { $0.localFeedImage }
+					completion(.found(feed: localFeedImages, timestamp: cache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [cacheStore, feedImageStore] in
+			do {
+				let cache = try cacheStore.existingCacheOrNewOne()
+				cache.feedImages = feed.map { feedImageStore.feedImage(from: $0) }
```

Comment:
Following from my previous comment, this mapping logic should also be extracted to one of the helpers. It's possible to create a function of type `([LocalFeedImage]) -> [ManagedFeedImage]` in `ManagedFeedImageStore`, and move the mapping logic there:
```suggestion
				cache.feedImages = feedImageStore.images(from: feed)
```